core/Pos/Core/Genesis/Types.hs

```diff
@@ -57,24 +84,72 @@ getTotalStake (BitcoinStakes _ st) = st
 getTotalStake RichPoorStakes {..} = unsafeIntegerToCoin $
     coinToInteger sdRichStake * fromIntegral sdRichmen +
     coinToInteger sdPoorStake * fromIntegral sdPoor
-getTotalStake (ExponentialStakes n) =
-    mkCoin $ sum $ map (2^) [0 .. n - 1]
+getTotalStake (ExponentialStakes n (fromIntegral . unsafeGetCoin -> mc)) =
+    mkCoin $ sum $ take (fromIntegral n) $ iterate (*2) mc
 getTotalStake (CustomStakes balances) =
     unsafeIntegerToCoin $ sumCoins balances
 
 -- | Distributions accompained by related addresses set (what to
 -- distribute and how).
 type AddrDistribution = ([Address], StakeDistribution)
 
+-- | Wrapper around weighted stakeholders map to be used in genesis
+-- core data.
+newtype GenesisWStakeholders = GenesisWStakeholders
+    { getGenesisWStakeholders :: HashMap StakeholderId Word16
+    } deriving (Show, Eq)
+
+instance Buildable GenesisWStakeholders where
+    build (GenesisWStakeholders m) =
+        bprint ("GenesisWStakeholders: "%listJson)
+               (map (sformat pairF) $ HM.toList m)
+
 -- | Hardcoded genesis data to generate utxo from.
 data GenesisCoreData = UnsafeGenesisCoreData
     { gcdAddrDistribution      :: ![AddrDistribution]
       -- ^ Address distribution. Determines utxo without boot
       -- stakeholders distribution (addresses and coins).
-    , gcdBootstrapStakeholders :: !(HashMap StakeholderId Word16)
+    , gcdBootstrapStakeholders :: !GenesisWStakeholders
       -- ^ Bootstrap era stakeholders, values are weights.
     } deriving (Show, Eq, Generic)
 
+-- | Calculates a minimum amount of coins user can set as an output in
+-- boot era.
+bootDustThreshold :: GenesisWStakeholders -> Coin
+bootDustThreshold (GenesisWStakeholders bootHolders) =
+    -- it's safe to use it here because weights are word16 and should
+    -- be really low in production, so this sum is not going to be
+    -- even more than 10-15 coins.
+    unsafeIntegerToCoin . sum $ map fromIntegral $ HM.elems bootHolders
+
+-- | Checks whether txOutDistribution matches the set of weighted boot
+-- stakeholders. Notice: it doesn't use actual txdistr type because
+-- it's defined in txp module above.
+bootRelatedDistr :: GenesisWStakeholders -> [(StakeholderId, Coin)] -> Bool
```

Comment:
It's just a consensus rule from txp. Just like any other consensus rule from there. And it's used only in one place. It fits there logically quite nicely, especially if you use actual txdistr type.